### PR TITLE
Update winit to version 22

### PR DESCRIPTION
**Pull request by yanchith**
A few things of note:

- The fullscreen option was removed, as we tend to use maximized instead.

- `imgui-winit-support` hasn't caught up with new winit changes, so it got
  replaced with our own, vendored and simplified version. This should make
  future upgrades easier.

- There is now a new `NewEvents` event, where we explicitly start timing a new
  frame and tell the input manager a new frame has started. In theory, the new
  approach should more correct because we now do event processing with correct
  time, but we probably didn't use time for event processing, so no bad behavior
  was exhibited previously.

- `winit::event::Event` no longer implements `Clone`, so we no longer can do
  event buffering (at least not easily).

- The old winit modifiers API is deprecated, so this commit replaces it with the
  new one.

Tested on:

- [x] Windows
  - ~Currently crashes when `tinyfiledialogs` tries to open a dialog. Might be an upstream problem in winit winapi backend. Stacktrace included below. Doesn't happen on other platforms.~ (fixed in winit 0.22.0)
- [x] macOS 
- [x] X11

Windows stacktrace:

```
thread 'main' panicked at 'already mutably borrowed: BorrowError', src\libcore\result.rs:1165:5
stack backtrace:
   0:     0x7ff7a6915899 - backtrace::backtrace::trace_unsynchronized
                               at C:\Users\VssAdministrator\.cargo\registry\src\github.com-1ecc6299db9ec823\backtrace-0.3.37\src\backtrace\mod.rs:66
   1:     0x7ff7a6915899 - std::sys_common::backtrace::_print_fmt
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\sys_common\backtrace.rs:76
   2:     0x7ff7a6915899 - std::sys_common::backtrace::_print::{{impl}}::fmt
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\sys_common\backtrace.rs:60
   3:     0x7ff7a692d38b - core::fmt::write
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libcore\fmt\mod.rs:1030
   4:     0x7ff7a6912504 - std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\io\mod.rs:1412
   5:     0x7ff7a69183b0 - std::sys_common::backtrace::_print
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\sys_common\backtrace.rs:64
   6:     0x7ff7a69183b0 - std::sys_common::backtrace::print
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\sys_common\backtrace.rs:49
   7:     0x7ff7a69183b0 - std::panicking::default_hook::{{closure}}
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:196
   8:     0x7ff7a6917fda - std::panicking::default_hook
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:210
   9:     0x7ff7a6918c08 - std::panicking::rust_panic_with_hook
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:473
  10:     0x7ff7a6918774 - std::panicking::continue_panic_fmt
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:380
  11:     0x7ff7a6918659 - std::panicking::rust_begin_panic
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:307
  12:     0x7ff7a6929019 - core::panicking::panic_fmt
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libcore\panicking.rs:85
  13:     0x7ff7a692911f - core::result::unwrap_failed
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libcore\result.rs:1165
  14:     0x7ff7a67eecbf - core::result::Result<core::cell::Ref<core::option::Option<winit::platform_impl::platform::event_loop::runner::EventLoopRunner<()>>>, core::cell::BorrowError>::expect<core::cell::Ref<core::option::Option<winit::platform_impl::platform::event_loop::runner::
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libcore\result.rs:960
  15:     0x7ff7a67fc79c - core::cell::RefCell<core::option::Option<winit::platform_impl::platform::event_loop::runner::EventLoopRunner<()>>>::borrow<core::option::Option<winit::platform_impl::platform::event_loop::runner::EventLoopRunner<()>>>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libcore\cell.rs:788
  16:     0x7ff7a680a788 - winit::platform_impl::platform::event_loop::runner::ELRShared<()>::send_event<()>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop\runner.rs:133
  17:     0x7ff7a5decdfb - winit::platform_impl::platform::event_loop::SubclassInput<()>::send_event<()>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop.rs:104
  18:     0x7ff7a5df0bee - winit::platform_impl::platform::event_loop::public_window_callback<()>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop.rs:1382
  19:     0x7ffe1201b762 - DefSubclassProc
  20:     0x7ffe1201b625 - DefSubclassProc
  21:     0x7ffe335f5b6d - CallWindowProcW
  22:     0x7ffe335f583e - CallWindowProcW
  23:     0x7ffe00553919 - glPushClientAttrib
  24:     0x7ffe335f5b6d - CallWindowProcW
  25:     0x7ffe335f573c - DispatchMessageW
  26:     0x7ffe33601ed3 - IsWindowVisible
  27:     0x7ffe3413fe14 - KiUserCallbackDispatcher
  28:     0x7ffe310b1ac4 - NtUserSetFocus
  29:     0x7ffe33155600 - GetFileTitleW
  30:     0x7ffe331b0f1c - PrintDlgA
  31:     0x7ffe331afb24 - PrintDlgA
  32:     0x7ffe331a7526 - PrintDlgA
  33:     0x7ffe331893d3 - Ssync_ANSI_UNICODE_Struct_For_WOW
  34:     0x7ffe33189b80 - GetOpenFileNameW
  35:     0x7ff7a69346d6 - tinyfd_openFileDialogW
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\tinyfiledialogs-3.3.6\libtinyfiledialogs\tinyfiledialogs.c:1626
  36:     0x7ff7a6936540 - openFileDialogWinGui8
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\tinyfiledialogs-3.3.6\libtinyfiledialogs\tinyfiledialogs.c:1701
  37:     0x7ff7a6932fe5 - tinyfd_openFileDialog
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\tinyfiledialogs-3.3.6\libtinyfiledialogs\tinyfiledialogs.c:3029
  38:     0x7ff7a5f9bce6 - tinyfiledialogs::open_file_dialog_impl
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\tinyfiledialogs-3.3.6\src\lib.rs:199
  39:     0x7ff7a5f9c20a - tinyfiledialogs::open_file_dialog
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\tinyfiledialogs-3.3.6\src\lib.rs:225
  40:     0x7ff7a5d70fc1 - hurban_selector::ui::file_input
                               at C:\Users\yanchith\workspace\subdgtl\HURBAN-Selector\src\ui.rs:1202
  41:     0x7ff7a5de78e2 - hurban_selector::ui::{{impl}}::draw_pipeline_window::{{closure}}
                               at C:\Users\yanchith\workspace\subdgtl\HURBAN-Selector\src\ui.rs:715
  42:     0x7ff7a5efd9ed - imgui::window::Window::build<closure-0>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\imgui-0.2.1\src\window\mod.rs:539
  43:     0x7ff7a5d6e8fe - hurban_selector::ui::UiFrame::draw_pipeline_window
                               at C:\Users\yanchith\workspace\subdgtl\HURBAN-Selector\src\ui.rs:517
  44:     0x7ff7a5f26aa2 - hurban_selector::init_and_run::{{closure}}
                               at C:\Users\yanchith\workspace\subdgtl\HURBAN-Selector\src\lib.rs:304
  45:     0x7ff7a5decca7 - winit::platform_impl::platform::event_loop::{{impl}}::run_return::{{closure}}<(),closure-0>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop.rs:207
  46:     0x7ff7a6825a53 - alloc::boxed::{{impl}}::call_mut<(winit::event::Event<()>, mut winit::event_loop::ControlFlow*),FnMut<(winit::event::Event<()>, mut winit::event_loop::ControlFlow*)>>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\liballoc\boxed.rs:929
  47:     0x7ff7a680d046 - winit::platform_impl::platform::event_loop::runner::{{impl}}::call_event_handler::{{closure}}<()>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop\runner.rs:563
  48:     0x7ff7a68321b6 - std::panic::{{impl}}::call_once<(),closure-0>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\panic.rs:315
  49:     0x7ff7a67ec462 - std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure-0>,()>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\panicking.rs:292
  50:     0x7ff7a691c552 - panic_unwind::__rust_maybe_catch_panic
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libpanic_unwind\lib.rs:80
  51:     0x7ff7a67ec2c3 - std::panicking::try<(),std::panic::AssertUnwindSafe<closure-0>>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\panicking.rs:271
  52:     0x7ff7a6833516 - std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure-0>,()>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\panic.rs:394
  53:     0x7ff7a680cdb0 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<()>::call_event_handler<()>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop\runner.rs:561
  54:     0x7ff7a680c678 - winit::platform_impl::platform::event_loop::runner::EventLoopRunner<()>::main_events_cleared<()>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop\runner.rs:456
  55:     0x7ff7a680b525 - winit::platform_impl::platform::event_loop::runner::ELRShared<()>::main_events_cleared<()>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop\runner.rs:203
  56:     0x7ff7a5dec7d4 - winit::platform_impl::platform::event_loop::EventLoop<()>::run_return<(),closure-0>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop.rs:239
  57:     0x7ff7a5decd16 - winit::platform_impl::platform::event_loop::EventLoop<()>::run<(),closure-0>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\platform_impl\windows\event_loop.rs:192
  58:     0x7ff7a5d457ed - winit::event_loop::EventLoop<()>::run<(),closure-0>
                               at C:\Users\yanchith\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.21.0\src\event_loop.rs:148
  59:     0x7ff7a5d44257 - hurban_selector::init_and_run
                               at C:\Users\yanchith\workspace\subdgtl\HURBAN-Selector\src\lib.rs:250
  60:     0x7ff7a5d4241c - hurban_selector::main
                               at C:\Users\yanchith\workspace\subdgtl\HURBAN-Selector\src\main.rs:92
  61:     0x7ff7a5d41f20 - std::rt::lang_start::{{closure}}<()>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\rt.rs:64
  62:     0x7ff7a69185b7 - std::rt::lang_start_internal::{{closure}}
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\rt.rs:49
  63:     0x7ff7a69185b7 - std::panicking::try::do_call<closure-0,i32>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:292
  64:     0x7ff7a691c552 - panic_unwind::__rust_maybe_catch_panic
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libpanic_unwind\lib.rs:80
  65:     0x7ff7a6918f42 - std::panicking::try
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panicking.rs:271
  66:     0x7ff7a6918f42 - std::panic::catch_unwind
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\panic.rs:394
  67:     0x7ff7a6918f42 - std::rt::lang_start_internal
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\/src\libstd\rt.rs:48
  68:     0x7ff7a5d41efb - std::rt::lang_start<()>
                               at /rustc/4560ea788cb760f0a34127156c78e2552949f734\src\libstd\rt.rs:64
  69:     0x7ff7a5d42440 - main
  70:     0x7ff7a6b97280 - invoke_main
                               at d:\agent\_work\5\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  71:     0x7ff7a6b97280 - __scrt_common_main_seh
                               at d:\agent\_work\5\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
  72:     0x7ffe33097bd4 - BaseThreadInitThunk
  73:     0x7ffe3410ced1 - RtlUserThreadStart
error: process didn't exit successfully: `target\debug\hurban_selector.exe` (exit code: 0xc0000409, STATUS_STACK_BUFFER_OVERRUN)
```

### CODE REVIEW

**Comment by yanchith** [comment]
Hmm, ok... Windows doesn't crash with winit 20, but macOS doesn't receive mouse events and X11 doesn't even start. I guess this will be blocked for a while.

**Comment by yanchith** [comment]
Ok, seems like winit 21 has less problems than winit 20, except for the `RefCell` double borrow bug on Windows. Likely culprit:

https://github.com/rust-windowing/winit/commit/6a330a2894873d29fbbfdeebfc1a215577213996

I'll bisect tomorrow.

**Comment by yanchith** [comment]
Alright. Updated to winit 0.22. Tested on X11 so far. Still need to test Windows and macOS.

**Comment by yanchith** [comment]
@janper @ondrowan Ready for review. I updated the description.

There is one minor issue on Windows still. The window does not start maximized. We maximize it after creation, but that has a small lag and is less pretty. I hope this is acceptable as a temporary workaround.

File: README.md [hunk 1]
```diff
@@ -66,9 +66,6 @@ Optionally configure one of the following:
 
 **HS_THEME**: Either `dark` or `funky`. Default is `dark`.
 
-**HS_FULLSCREEN**: Run (1) or don't run (0) in fullscreen. Default is
-  no fullscreen.
-
```
Comment (janper) [inline README.md:71]: When did we decide to remove this option? I recall we agreed to leave it.

**Review by janper — APPROVED** [review]
The maximized startup looks fairly similar to the previous version - can't tell any difference actually.
BTW, when did we decide to loose the fullscreen support?

File: README.md (hunk 1 above)
Comment (yanchith) [reply README.md:71]: Maybe I misunderstood the outcome of our conversation. Do you want me to add it back?

**Comment by yanchith** [comment]
> The maximized startup looks fairly similar to the previous version

Glad I didn't make the situation worse :smile: The startup is much nicer on macOS and X11, but what can we do...

File: README.md (hunk 1 above)
Comment (janper) [reply README.md:71]: No, but let's hold a moment of silence for yet another feature removed from this software.

**Pull request merged**